src/FakeItEasy/SelfInitializedFakes/MethodInfoConverter.cs

```diff
@@ -0,0 +1,118 @@
+﻿namespace FakeItEasy.SelfInitializedFakes
+{
+    using System;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Reflection;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// A custom converter for MethodInfo.
+    /// </summary>
+    public class MethodInfoConverter : JsonConverter
+    {
+        /// <summary>
+        /// Determines whether this instance can convert the specified object type.
+        /// </summary>
+        /// <param name="objectType">Type of the object.</param>
+        /// <returns>true if this instance can convert the specified object type; otherwise, false.</returns>
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType == typeof(MethodInfo);
+        }
+
+        /// <summary>
+        /// Reads the JSON representation of the object.
+        /// </summary>
+        /// <param name="reader">The <see cref="Newtonsoft.Json.JsonReader"/> to read from.</param>
+        /// <param name="objectType">Type of the object.</param>
+        /// <param name="existingValue">The existing value of object being read.</param>
+        /// <param name="serializer">The calling serializer.</param>
+        /// <returns>The object value.</returns>
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            reader.Read();
+            Debug.Assert(
+                (string)reader.Value == nameof(MethodInfoData.QualifiedTypename),
+                "Unexpected Json property name {0}. Expected: {1}.",
+                (string)reader.Value,
+                nameof(MethodInfoData.QualifiedTypename));

```

Comment:
The point may eventually become moot, if we drop serialization altogether, but if the code stays, we'll eventually need a better approach than `Debug.Assert` to indicate an error.